### PR TITLE
Update fmt function for Key

### PR DESCRIPTION
**Pull request by shinghim**
Currently, the `Display` function for `Key` is implemented to display it in the following format: 
```rust
"type: {:#x}, key: {:x}"
```

From #2869 discussions, the plan in the future is to call through to `to_string` from the `to_hex` function, so `to_string` should return a pure-hex representation of the type. For `Key`, this would mean removing the `type: ` and `key: ` labels in the string being returned in `Display`

### CODE REVIEW

**Review by Kixunil — CHANGES_REQUESTED** [review]
It's the other way around. The code is correct but the comment claiming there's a key type in the `Vec` is not. It's just the key itself - look at the deserialization code. So just change/delete the comment.

**Comment by jamillambert** [comment]
The comment ```/// `<key> := <keylen> <keytype> <keydata>` ``` was added in #988 to give the format of a PSBT key as described in [BIP 174](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki), but it would be better if it described how it is represented here in plain english, similarly for the `Pair` struct.

There is also a comment at the bottom of the file which suggests the type is contained in the `key: Vec`.  It should say "if the key `type_value` is not `0xFC`" or something similar.

```
    /// Returns [`Error::InvalidProprietaryKey`] if `key` does not start with `0xFC` byte.
    fn try_from(key: Key) -> Result<Self, Self::Error> {
        if key.type_value != 0xFC {
            return Err(Error::InvalidProprietaryKey);
        }
```

**Comment by tcharding** [comment]
Yes, what @jamillambert said. FWIW I think these lines from the bip add value but perhaps move them all onto the structs not the fields which is a bit confusing.

Probably better to use the term 'key data' in what is now `    /// The key itself in raw byte form.` same as we do on the `value` field of `Pair`.

Thanks!

**Comment by shinghim** [comment]
ah, makes sense. I agree the documentation is useful but would be better suited in for the struct docs. I'll update that as well as the variable names

File: bitcoin/src/psbt/raw.rs [hunk 1]
```diff
@@ -68,7 +67,10 @@ where
 
 impl fmt::Display for Key {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "type: {:#x}, key: {:x}", self.type_value, self.key.as_hex())
+        let mut key_bytes = vec![0u8; self.key.len() + 1];
+        key_bytes[0] = self.type_value;
+        key_bytes[1..].copy_from_slice(&self.key);
+        write!(f, "{}", key_bytes.as_hex())
```
[the remark this answers is not in the document]
Comment (shinghim) [reply bitcoin/src/psbt/raw.rs:73]: This will display just the hex key when we do the call-through to `to_string` from `to_hex`. Makes sense to display one after another, I'll update it to do that

[the remark this answers is not in the document]
Comment (Kixunil) [reply bitcoin/src/psbt/raw.rs]: I know what it does. I'm asking why did you change it from human readable to human unreadable.

[the remark this answers is not in the document]
Comment (shinghim) [reply bitcoin/src/psbt/raw.rs]: This is in preparation for #2869, where `to_hex` will do a call-through to `to_string`. It makes sense to have something human-readable for `to_string`, but I dont think what's currently implemented will be what is expected when calling `to_hex`. You also mentioned [something similar about returning pure hex](https://github.com/rust-bitcoin/rust-bitcoin/issues/2869#issuecomment-2206381458) - not sure if this is what you meant as well. If it makes more sense to not have the call-through to `to_string`, that's also fine; we can update #2869 to reflect that

**Comment by tcharding** [comment]
I see this is draft still but from the commit log I don't know what this PR is trying to do " Fix fmt function and documentation for Key", what's broken, why are we fixing it? Can you elaborate in the PR, and perhaps consider splitting the docs changes into a separate patch if they are just structural changes not content changes. Thanks

**Comment by shinghim** [comment]
Updated the PR description/comment - let me know if it clears it up. I do think it makes sense to split up the doc changes and the changes to `Display`

**Comment by shinghim** [comment]
if it's not necessary please feel free to just close this

**Comment by tcharding** [comment]
Ah, thanks man, now I see whats going on. We definitely should keep going with this if you have the patience.

**Comment by shinghim** [comment]
Of course! I'd like to see this through

File: bitcoin/src/psbt/raw.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kixunil) [reply bitcoin/src/psbt/raw.rs]: > If it makes more sense to not have the call-through to `to_string`

Indeed, it does, I think. Similar to how `Script` displays as ASM and has a `to_hex` method.

[the remark this answers is not in the document]
Comment (shinghim) [reply bitcoin/src/psbt/raw.rs]: yep, though this type falls under the exception [you mentioned](https://github.com/rust-bitcoin/rust-bitcoin/pull/2318#discussion_r1453062815):

> Only for types with unknown encoded lengths (e.g. all consensus types containing a `Vec`)

I opened this issue to prep for that call through since `Key` has the `Vec` field of `key`. I'm not sure how often `to_string` is being used for `Key`, but if `to_hex` for other types are call throughs to `to_string`, I'd lean more in that direction here as well for consistency

[the remark this answers is not in the document]
Comment (Kixunil) [reply bitcoin/src/psbt/raw.rs]: I can't remember why I said it. We can trivially pre-allocate here - `String::with_capacity(self.key_len * 2 + 2)`.

[the remark this answers is not in the document]
Comment (shinghim) [reply bitcoin/src/psbt/raw.rs]: Is this what's going to be returned for `Display`? Why do we need `self.key_len * 2 + 2`? is there a scenario where we would have more data than `<key> := <keylen> <keytype> <keydata>`?

[the remark this answers is not in the document]
Comment (Kixunil) [reply bitcoin/src/psbt/raw.rs]: No, that's in `to_hex`. `self.key_len * 2 + 2` is literally the length of the resulting string.

**Comment by shinghim** [comment]
Closing this since it sounds like we will keep `Display` the same and just create an independent `to_hex` function

**Comment by Kixunil** [comment]
OK. Updating the doc would still be useful though.

**Comment by shinghim** [comment]
sure, will put up a PR to update the doc
